extensions/ql-vscode/src/queries-panel/query-discovery.ts

```diff
@@ -0,0 +1,183 @@
+import * as vscode from "vscode";
+import { dirname, basename, join, normalize, relative } from "path";
+import { Discovery } from "../common/discovery";
+import { CodeQLCliServer } from "../codeql-cli/cli";
+import { pathExists } from "fs-extra";
+
+/**
+ * A node in the tree of queries. This will be either a `QueryDirectory` or a `QueryFile`.
+ */
+export abstract class QueryNode {
+  constructor(private _path: string, private _name: string) {}
+
+  public get path(): string {
+    return this._path;
+  }
+
+  public get name(): string {
+    return this._name;
+  }
+
+  public abstract get children(): readonly QueryNode[];
+
+  public abstract finish(): void;
+}
+
+/**
+ * A directory containing one or more query files or other query directories.
+ */
+export class QueryDirectory extends QueryNode {
+  constructor(
+    _path: string,
+    _name: string,
+    private _children: QueryNode[] = [],
+  ) {
+    super(_path, _name);
+  }
+
+  public get children(): readonly QueryNode[] {
+    return this._children;
+  }
+
+  public addChild(child: QueryNode): void {
+    this._children.push(child);
+  }
+
+  public createDirectory(relativePath: string): QueryDirectory {
+    const dirName = dirname(relativePath);
+    if (dirName === ".") {
+      return this.createChildDirectory(relativePath);
+    } else {
+      const parent = this.createDirectory(dirName);
+      return parent.createDirectory(basename(relativePath));
+    }
+  }
+
+  public finish(): void {
+    // remove empty directories
+    this._children.filter(
+      (child) => child instanceof QueryFile || child.children.length > 0,
+    );
+    this._children.sort((a, b) =>
+      a.name.localeCompare(b.name, vscode.env.language),
+    );
+    this._children.forEach((child, i) => {
+      child.finish();
+      if (
+        child.children?.length === 1 &&
+        child.children[0] instanceof QueryDirectory
+      ) {
+        // collapse children
+        const replacement = new QueryDirectory(
+          child.children[0].path,
+          `${child.name} / ${child.children[0].name}`,
+          Array.from(child.children[0].children),
+        );
+        this._children[i] = replacement;
+      }
+    });
```

Comment:
We were planning to do this "directory collapsing" in a follow-up issue, but this seems to do the job... (Currently copied over from [here](https://github.com/github/vscode-codeql/blob/d0ca885e805ebf3d3b45435cc6a5522494e78e06/extensions/ql-vscode/src/query-testing/qltest-discovery.ts#L64-L85)—perhaps this could be a shared helper function 🤔 )